proposals/p2665.md

```diff
@@ -0,0 +1,209 @@
+
+-   [Abstract](#abstract)
+-   [Problem](#problem)
+-   [Background](#background)
+    -   [Discussion in Carbon](#discussion-in-carbon)
+    -   [In other languages](#in-other-languages)
+        -   [Requiring semicolons](#requiring-semicolons)
+        -   [Optional semicolons](#optional-semicolons)
+-   [Proposal](#proposal)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+    -   [Optional semicolons](#optional-semicolons-1)
+
+<!-- tocstop -->
+
+## Abstract
+
+Statements, declarations, and definitions will terminate with either a semicolon
+(`;`) or a close curly brace (`}`). Semicolons are never optional.
+
+For example, with a semicolon, `x = x + 2;` or `class C;`. With a close curly
+brace, `for ( ... ) { ... }`, or `class C { ...}`.
+
+This does not affect any approved proposal; rather, it makes an important
+assumption explicit.
+
+## Problem
+
+Statements need some system for separation. There are two main options for this:
+
+1. Require semicolons to terminate statements.
+2. Automatically determine where statements terminate.
+    - Some languages, such as Python, define a syntax where a newline terminates
+      statements.
+    - Other languages, such as Javascript, require semicolons but define rules
+      for semicolon insertion.
+
+Although Carbon's design currently assumes semicolons are required, it hasn't
+been directly addressed by a proposal.
+
+## Background
+
+### Discussion in Carbon
+
+This was discussed on leads issue
+[#1924: Semicolon](https://github.com/carbon-language/carbon-lang/issues/1924).
+Some rationale is provided there, stemming from discussion
+[#1739: Semicolon](https://github.com/carbon-language/carbon-lang/discussions/1739).
+
+### In other languages
+
+#### Requiring semicolons
+
+In C++, C#, and Java, semicolons are always required.
+
+In Rust, semicolons are generally required, but may be omitted for an
+[implicit return](https://doc.rust-lang.org/std/keyword.return.html).
+
+#### Optional semicolons
+
+In Python, a line is a
+[simple statement](https://docs.python.org/3/reference/simple_stmts.html), and
+parentheses are an idiomatic way to create multi-line statements. Semicolons may
+be used to explicitly separate statements. For example:
+
+```python
+value = (
+  "text"
+)
+a = 1; b = 2; c = 3
+```
+
+Swift takes a different approach, where
+[statement separation is part of the grammar](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/statements).
+This allows some statements to wrap lines. However, Swift still disallows
+multiple statements on the same line unless a semicolon is present. For example:
+
+```swift
+// One statement in Swift, but two in Python and Kotlin.
+var x = 1
+      + 1
+// Two statements in Swift because of whitespace sensitivity.
+var x = 1
+      +1
+```
+
+Kotlin makes
+[newlines usable instead of semicolons in code blocks](https://kotlinlang.org/spec/statements.html#code-blocks).
```

Comment:
I think it'd be more useful to say:

Kotlin permits a newline to be used to terminate statements instead of a semicolon. Kotlin's grammar [explicitly enumerates](https://kotlinlang.org/spec/syntax-and-grammar.html) all the places where newlines can appear (see mentions of `NL` in the grammar), and doesn't allow newlines in places where they would introduce ambiguity.
```kotlin
// This is unambiguously parsed as two statements, because
// a newline is not permitted before a `+` operator.
var x = 1
+ 1
```